### PR TITLE
Implement storager

### PR DESCRIPTION
**Pull request by junaire**
This is the first PR of my work. It intends to finish the tasks below:
- [x] `NewStorager`
- [x] `formatError`

### General tracking issue #6

### CODE REVIEW

File: errors.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package gdrive
+
+import (
+	"github.com/beyondstorage/go-storage/v4/services"
+)
+
+var (
+	ErrRateLimitExceeded  = services.NewErrorCode("Rate limit exceeded")
+	ErrBackEnd            = services.NewErrorCode("Backend Error")
```
Comment (Xuanwo) [inline errors.go]: Please use `ErrServiceInternal` instead.

Ref:

- https://github.com/beyondstorage/specs/blob/master/rfcs/127-add-serviceinternel-and-requestthrottled-errors.md
- https://github.com/beyondstorage/go-storage/issues/612

File: errors.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+	ErrInvalidCredentials = services.NewErrorCode("Invalid credentials")
```
Comment (Xuanwo) [inline errors.go]: Please use errors provided by `credentials` package.

File: utils.go [hunk 3]
```diff
@@ -1,11 +1,24 @@
...
+	"google.golang.org/api/googleapi"
+	"google.golang.org/api/option"
 )
 
 // Storage is the example client.
 type Storage struct {
+	name    string
+	workDir string
+	service *drive.Service
+	//pairPolicy   types.PairPolicy
```
Comment (Xuanwo) [inline utils.go]: Please remove useless code.

File: errors.go [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package gdrive
+
+import (
+	"github.com/beyondstorage/go-storage/v4/services"
+)
+
+var (
+	ErrRateLimitExceeded  = services.NewErrorCode("Rate limit exceeded")
```
Comment (Xuanwo) [inline errors.go]: Please use `ErrRequestThrottled` instead.

**Comment by Xuanwo** [comment]
It's OK for me to split our project into several PRs. Maybe we can set up a roadmap for it?

File: errors.go (hunk 1 above)
Comment (junaire) [reply errors.go]: Is this proposal has been implemented? I just can't find it in `go-storage/services/error.go` Can we pass the tests if I use it?

**Comment by junaire** [comment]
> It's OK for me to split our project into several PRs. Maybe we can set up a roadmap for it?

My original thought is that open one PR and add more commits into it. You can review my work and merge the PR when our work is done. But it is OK if you prefer to split one PR into several ones and merge them one by one. As for the roadmap, I would like to open a new issue, and list all the work I need to finish. Finally we can close it when all of PRs get merged. What's your idea?

**Comment by Xuanwo** [comment]
>  As for the roadmap, I would like to open a new issue, and list all the work I need to finish. Finally we can close it when all of PRs get merged.

Great! Looks good to me.

File: errors.go (hunk 1 above)
Comment (Xuanwo) [reply errors.go]: > Is this proposal has been implemented? I just can't find it in `go-storage/services/error.go` Can we pass the tests if I use it?

ping @xxchan to get the plan for implementations of GSP-127.

As GSP-127 is not implemented, return error like `erros.New("xxx")` is OK to me. It's better than define a public error too early.

Comment (junaire) [reply errors.go]: How about just simply use these unimplemented errors, we can left the failed tests and merge the PR when GSP-127 is done.

Comment (Xuanwo) [reply errors.go]: OK.

**Comment by junaire** [comment]
It doesn't look too difficult to implement GSP-127, do you mind if I open a PR? @Xuanwo If @xxchan is too busy?

**Comment by Xuanwo** [comment]
> It doesn't look too difficult to implement GSP-127, do you mind if I open a PR? @Xuanwo If @xxchan is too busy?

This is so great! Please feel free to send a PR for GSP-127.

**Comment by junaire** [comment]
Hey, I'm back! I finally finished my final exams and now I can work on the project full time! ;-) Besides, is this PR ready to be merged? Since we have finished GSP-127, it should pass the tests successfully. However, I don't know why I can't make it on my local machine. Is it because I should update the 3rd party packages manually? I have tried `go get -u all`, but it didn't work. :-( I'm sorry but I'm a newbie to the Go, and I'm little confused with the package management.

**Comment by xxchan** [comment]
You are using go-storage v4.2.0. Your GSP-127 implementation is on `master`, not released yet. 

`go get -u` will try to use newer minor or patch **releases** when available.

Try `go get github.com/beyondstorage/go-storage/v4@master`

You can read:
https://golang.org/cmd/go/#hdr-Add_dependencies_to_current_module_and_install_them
https://xuanwo.io/2019/05/27/go-modules/

**Review by Xuanwo — APPROVED** [review]

**Pull request merged**